### PR TITLE
Rewrite `DefaultTextArgumentSplicer` and add new `FromCode` attribute

### PR DESCRIPTION
**Pull request by OoLunar**
# Summary
This PR turned into something much larger than I ever intended it to be:
- `DefaultTextArgumentSplicer` has been completely rewritten instead of using the old CommandsNext text argument parser.
- `DSharpPlus.Tests` has switched away from MSTest and over to NUnit. The main motivator being that MSTest required an external database or an XML file in order to use `TestDataSource`. This is not the case with NUnit as NUnit will happily use other properties/methods to provide different arguments to our test methods.
- I noticed that the `DSharpPlus.Tests` file architecture didn't match the namespaces. This is no longer the case.
- A new `FromCode` attribute which strips codeblocks from user input. 

# Notes
All changes have been tested. New test cases are encouraged.

### CODE REVIEW

**Review by Plerx2493 — APPROVED** [review]
LGTM

File: Directory.Packages.props [hunk 1]
```diff
@@ -2,20 +2,21 @@
...
-    <PackageVersion Include="MSTest.TestAdapter" Version="3.2.2" />
-    <PackageVersion Include="MSTest.TestFramework" Version="3.2.2" />
     <PackageVersion Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.4" />
-    <PackageVersion Include="Microsoft.CodeAnalysis.CSharp" Version="4.7.0" />
     <PackageVersion Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="4.6.0-1.final" />
+    <PackageVersion Include="Microsoft.CodeAnalysis.CSharp" Version="4.7.0" />
     <PackageVersion Include="Microsoft.Extensions.Caching.Memory" Version="8.0.0" />
-    <PackageVersion Include="Microsoft.Extensions.Configuration" Version="7.0.0" />
     <PackageVersion Include="Microsoft.Extensions.Configuration.CommandLine" Version="7.0.0" />
+    <PackageVersion Include="Microsoft.Extensions.Configuration" Version="7.0.0" />
```
Comment (akiraveliara) [inline Directory.Packages.props:10]: these two are misordered now

File: Directory.Packages.props [hunk 2]
```diff
@@ -2,20 +2,21 @@
...
     <PackageVersion Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0" />
     <PackageVersion Include="Microsoft.Extensions.DependencyModel" Version="7.0.0" />
-    <PackageVersion Include="Microsoft.Extensions.Logging" Version="8.0.0" />
     <PackageVersion Include="Microsoft.Extensions.Logging.Abstractions" Version="8.0.0" />
-    <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
+    <PackageVersion Include="Microsoft.Extensions.Logging" Version="8.0.0" />
+    <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
     <PackageVersion Include="Newtonsoft.Json" Version="13.0.3" />
+    <PackageVersion Include="NUnit.Analyzers" Version="3.10.0" />
+    <PackageVersion Include="NUnit" Version="4.0.1" />
```
Comment (akiraveliara) [inline Directory.Packages.props:18]: as are several here

**Review by akiraveliara — CHANGES_REQUESTED** [review]

File: DSharpPlus.Commands/ContextChecks/FromCodeAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+namespace DSharpPlus.Commands.ContextChecks;
+
+using System;
+
+[AttributeUsage(AttributeTargets.Parameter)]
+public sealed class FromCodeAttribute(CodeType codeType = CodeType.All) : Attribute
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/ContextChecks/FromCodeAttribute.cs:6]: Where else should it go?

[the remark this answers is not in the document]
Comment (akiraveliara) [reply DSharpPlus.Commands/ContextChecks/FromCodeAttribute.cs:6]: somewhere logically related to text command parameter shapes. Processors.Text.Parameters/ParameterModel/ParameterShapes/whatever?

File: DSharpPlus.Tests/Commands/Cases/UserInput.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+using System.Collections.Generic;
+using NUnit.Framework;
+
+namespace DSharpPlus.Tests.Commands.Cases;
+
+public static class UserInput
+{
+    public static readonly List<TestCaseData> ExpectedNormal =
+    [
+        new TestCaseData("Hello", 1, new[] { "Hello" }),
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Tests/Commands/Cases/UserInput.cs:10]: > `Argument 2: cannot convert from 'collection expressions' to 'object'`

File: DSharpPlus.Commands/ContextChecks/FromCodeAttribute.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/ContextChecks/FromCodeAttribute.cs:6]: This attribute isn't specific to text commands

[the remark this answers is not in the document]
Comment (akiraveliara) [reply DSharpPlus.Commands/ContextChecks/FromCodeAttribute.cs:6]: what other commands can use code blocks? - but anyway, it sure isn't a check, so

File: DSharpPlus.Commands/Processors/TextCommands/Parsing/DefaultTextArgumentSplicer.cs [hunk 5]
```diff
@@ -1,137 +1,119 @@
...
         char[] quoteCharacters = extension.GetProcessor<TextCommandProcessor>().Configuration.QuoteCharacters;
-        bool inBacktick = false;
-        bool inTripleBacktick = false;
-        bool inQuote = false;
-        bool inEscape = false;
-        List<int> removeIndices = new(text.Length - startAt);
-
-        int i = startAt;
-        for (; i < text.Length; i++)
+        for (i = startAt; i < textSpan.Length; i++)
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/Processors/TextCommands/Parsing/DefaultTextArgumentSplicer.cs:33]: You're more than welcome to improve upon it

File: DSharpPlus.Commands/ContextChecks/FromCodeAttribute.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/ContextChecks/FromCodeAttribute.cs]: Slash commands via text input. They've supported multiline arguments for awhile now

File: DSharpPlus.Commands/Converters/StringConverter.cs [hunk 6]
```diff
@@ -16,14 +19,59 @@ public class StringConverter : ISlashArgumentConverter<string>, ITextArgumentCon
...
+        }
+    }
+
+    private static bool TryGetCodeBlock(string input, CodeType expectedCodeType, [NotNullWhen(true)] out string? code)
+    {
+        code = null;
+        if (input.Length > 6 && input.StartsWith("```") && input.EndsWith("```") && expectedCodeType.HasFlag(CodeType.Codeblock))
+        {
+            int newline = input.IndexOf('\n');
+            code = newline == -1 ? input[3..^3] : input[newline..^3];
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/Converters/StringConverter.cs]: Never doubt me again.

File: Directory.Packages.props [hunk 7]
```diff
@@ -14,8 +12,11 @@
     <PackageVersion Include="Microsoft.Extensions.DependencyModel" Version="7.0.0" />
     <PackageVersion Include="Microsoft.Extensions.Logging" Version="8.0.0" />
     <PackageVersion Include="Microsoft.Extensions.Logging.Abstractions" Version="8.0.0" />
-    <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
+    <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
     <PackageVersion Include="Newtonsoft.Json" Version="13.0.3" />
+    <PackageVersion Include="NUnit.Analyzers" Version="3.10.0" />
+    <PackageVersion Include="NUnit" Version="4.0.1" />
```
Comment (akiraveliara) [inline Directory.Packages.props:18]: why are we downgrading, and NUnit.Analyzers should be below NUnit

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/FromCodeAttribute.LanguageList.template [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// <auto-generated/>
+// Last modified on {{Date}}
+
+using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+
+namespace DSharpPlus.Commands.ContextChecks;
+
+public sealed partial class FromCodeAttribute : Attribute
```
Comment (akiraveliara) [inline tools/DSharpPlus.Tools.CodeBlockLanguageListGen/FromCodeAttribute.LanguageList.template:10]: just `partial class` so the other modifies can be changed in code

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+namespace DSharpPlus.Tools.CodeBlockLanguageListGen;
+
+public static class Program
+{
+    private static readonly string TemplateFile;
+
+    static Program()
+    {
+        // Load the template file
+        // The filename could probably be hardcoded.
```
Comment (akiraveliara) [inline tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs:20]: do that.

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            if (i != languages.Count - 1)
+            {
+                languageListNode.Append(',');
+            }
+
+            languageListNode.Append('\n');
+        }
+        languageListNode.Append("    }.ToFrozenSet()");
+
+        File.WriteAllText($"{Environment.CurrentDirectory}/DSharpPlus.Commands/ContextChecks/FromCode/FromCodeAttribute.LanguageList.cs", TemplateFile
```
Comment (akiraveliara) [inline tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs:54]: `"./DSharpPlus.Commands/ContextChecks..."`

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        throw new InvalidOperationException("Failed to load the template file.");
+    }
+
+    public static async Task Main()
+    {
+        // Grab the list of languages and their aliases
+        IReadOnlyList<string> languages = await GetLanguagesAsync();
+        StringBuilder languageListNode = new("new List<string>()\n");
```
Comment (akiraveliara) [inline tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs:40]: can we use a collection expression there?

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+        File.WriteAllText($"{Environment.CurrentDirectory}/DSharpPlus.Commands/ContextChecks/FromCode/FromCodeAttribute.LanguageList.cs", TemplateFile
+            .Replace("{{Date}}", DateTimeOffset.UtcNow.ToString("F", CultureInfo.InvariantCulture))
+            .Replace("{{CodeBlockLanguages}}", languageListNode.ToString()
+        ));
+    }
+
+    public static async ValueTask<IReadOnlyList<string>> GetLanguagesAsync()
+    {
+        ProcessStartInfo psi = new("node", "tools/get-highlight-languages")
```
Comment (akiraveliara) [inline tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs:62]: this creates a dev-time dependency on node. is this really something we want?

**Review by akiraveliara — CHANGES_REQUESTED** [review]

File: Directory.Packages.props (hunk 7 above)
Comment (OoLunar) [reply Directory.Packages.props:18]: Not intentional, unsure how that happened.

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs (hunk 11 above)
Comment (OoLunar) [reply tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs:40]: No. Collection expressions can only be used where `new()` can be used. Neither `new()` or `[]` know to create a list of strings

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs (hunk 12 above)
Comment (OoLunar) [reply tools/DSharpPlus.Tools.CodeBlockLanguageListGen/Program.cs:62]: Node is always installed on the GitHub runners. I can create a GitHub Action that opens up a PR if for whatever reason, none of the maintainers or contributors have Node installed.

Also it's the most accurate way to get a list of up-to-date languages that Discord supports.

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/FromCodeAttribute.LanguageList.template (hunk 8 above)
Comment (OoLunar) [reply tools/DSharpPlus.Tools.CodeBlockLanguageListGen/FromCodeAttribute.LanguageList.template:10]: No modifications can be made to the behavior as it's the argument converter that's doing the work. Remember how you didn't want context checks to be able to mutate parameter values?

File: DSharpPlus.Commands/Converters/StringConverter.cs [hunk 13]
```diff
@@ -16,14 +19,81 @@ public class StringConverter : ISlashArgumentConverter<string>, ITextArgumentCon
...
+        code = null;
+        if (input.Length > 6 && input.StartsWith("```") && input.EndsWith("```") && expectedCodeType.HasFlag(CodeType.Codeblock))
+        {
+            for (int i = 0; i < input.Length; i++)
+            {
+                // Read until we hit either a newline or a space.
+                if (!char.IsWhiteSpace(input[i]))
+                {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/Converters/StringConverter.cs:69]: `char` does not expose the whitespace characters. Should we hardcode `\n` and ` `?

File: tools/DSharpPlus.Tools.CodeBlockLanguageListGen/FromCodeAttribute.LanguageList.template (hunk 8 above)
Comment (akiraveliara) [reply tools/DSharpPlus.Tools.CodeBlockLanguageListGen/FromCodeAttribute.LanguageList.template:10]: the modifiers, not the behaviour

Comment (akiraveliara) [reply tools/DSharpPlus.Tools.CodeBlockLanguageListGen/FromCodeAttribute.LanguageList.template:10]: say, you want to remove sealed for whatever reason. that now needs to be changed in two places and the tool needs to re-run

File: DSharpPlus.Commands/Converters/StringConverter.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (akiraveliara) [reply DSharpPlus.Commands/Converters/StringConverter.cs:69]: probably

**Review by akiraveliara — APPROVED** [review]

**Pull request merged**